File: ru.kmorozov.gbd.db/src/main/java/ru/kmorozov/db/core/logic/model/book/google/GooglePageInfo.java
package ru.kmorozov.db.core.logic.model.book.google;

import com.google.gson.annotations.SerializedName;
import org.apache.commons.lang3.StringUtils;
import ru.kmorozov.gbd.core.logic.model.book.base.AbstractPage;

import java.io.Serializable;

import static ru.kmorozov.gbd.core.config.constants.GoogleConstants.*;

/**
 * Created by km on 21.11.2015.
 */

public class GooglePageInfo extends AbstractPage implements Serializable, Comparable<GooglePageInfo> {

    @SerializedName("pid")
    private String pid;
    @SerializedName("flags")
    private String flags;
    @SerializedName("title")
    private String title;
    private String src;
    @SerializedName("uf")
    private String uf;
    private transient String sig;

    @SerializedName("order")
    private int order;
    @SerializedName("h")
    private int h;
    private int width;
    @SerializedName("links")
    private Object links;

    private String prefix;
    private int pageNum = -1;

    private transient boolean gapPage;

    public GooglePageInfo() {
    }

    // Создание страниц для заполнения разрыва
    public GooglePageInfo(String pid, int order) {
        this.pid = pid;
        this.order = order;

        this.gapPage = true;
    }

    @Override
    public String getPid() {
        return this.pid;
    }

    @Override
    public Integer getOrder() {
        return this.order;
    }

    public String getSrc() {
        return this.src;
    }

    public void setSrc(String src) {
        this.src = src;
    }

    public void setWidth(int width) {
        this.width = width;
    }

    public CharSequence getSig() {
        return null == this.src ? null : null == this.sig ? this.sig = this.src.substring(this.src.indexOf("sig=") + 4) : this.sig;
    }

    public boolean isGapPage() {
        return this.gapPage;
    }

    public String getImqRqUrl(CharSequence bookId, String urlTemplate, int width) {
        return urlTemplate.replace(BOOK_ID_PLACEHOLDER, bookId) + IMG_REQUEST_TEMPLATE.replace(RQ_PG_PLACEHOLDER, this.pid)
                                                                                      .replace(RQ_SIG_PLACEHOLDER, this.getSig())
                                                                                      .replace(RQ_WIDTH_PLACEHOLDER, String
                                                                                              .valueOf(width));
    }

    public String getPrefix() {
        if (StringUtils.isEmpty(this.prefix)) this.parsePageNum();

        return this.prefix;
    }

    public int getPageNum() {
        return this.pageNum;
    }

    private void parsePageNum() {
        boolean numFound = false;
        this.prefix = "";
        StringBuilder strNum = new StringBuilder();
        for (int i = this.pid.length() - 1; 0 <= i; i--) {
            char ch = this.pid.charAt(i);
            if (!numFound && Character.isDigit(ch)) strNum.insert(0, ch);
            else if (!numFound && !Character.isDigit(ch)) {
                numFound = true;
                this.prefix = ch + this.prefix;
            }
            else this.prefix = ch + this.prefix;
        }

        try {
            this.pageNum = Integer.parseInt(strNum.toString());
        } catch (NumberFormatException ignored) {
        }
    }

    @Override
    public int compareTo(GooglePageInfo anotherPage) {
        return getOrder().compareTo(anotherPage.getOrder());
    }

    @Override
    public String getImgUrl() {
        throw new UnsupportedOperationException();
    }
}

File: ru.kmorozov.gbd.db/src/main/java/ru/kmorozov/db/core/logic/model/book/google/GooglePageInfo.kt
package ru.kmorozov.db.core.logic.model.book.google

import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName
import org.apache.commons.lang3.StringUtils
import ru.kmorozov.gbd.core.logic.model.book.base.AbstractPage
import ru.kmorozov.gbd.core.logic.model.book.base.IPage
import java.io.Serializable
import java.util.concurrent.ConcurrentHashMap

/**
 * Created by km on 21.11.2015.
 */

class GooglePageInfo : AbstractPage, Serializable, Comparable<IPage> {
    override val imgUrl: String
        get() = TODO("not implemented") //To change initializer of created properties use File | Settings | File Templates.

    @Expose
    @SerializedName("pid")
    override var pid: String = ""

    @SerializedName("flags")
    private val flags: String? = null

    @Expose
    @SerializedName("title")
    private val title: String? = null

    @Expose
    var src: String? = null
        private set

    @SerializedName("uf")
    private val uf: String? = null

    @Transient
    public var sigs: MutableSet<String> = ConcurrentHashMap.newKeySet()
        private set

    fun addSrc(_src: String): Boolean {
        src = _src
        val idx = _src.indexOf("sig=")

        return idx >= 0 && sigs.add(_src.substring(idx + 4))
    }

    @Expose
    @SerializedName("order")
    override var order: Int = 0

    @SerializedName("h")
    private val h: Int = 0
    private val width: Int = 0

    @SerializedName("links")
    private val links: Any? = null

    var prefix: String = ""
        get() = if (StringUtils.isEmpty(field)) parsePageNum() else field

    var pageNum = -1
        private set

    @Transient
    private var gapPage: Boolean = false

    constructor() {}

    // Создание страниц для заполнения разрыва
    constructor(pid: String, order: Int) {
        this.pid = pid
        this.order = order

        gapPage = true
    }

    private fun parsePageNum(): String {
        var numFound = false
        var _prefix = ""
        val strNum = StringBuilder()
        var i = pid.length - 1
        while (0 <= i) {
            val ch = pid[i]
            if (!numFound && Character.isDigit(ch))
                strNum.insert(0, ch)
            else if (!numFound && !Character.isDigit(ch)) {
                numFound = true
                _prefix = ch + _prefix
            } else
                _prefix = ch + _prefix
            i--
        }

        pageNum = Integer.parseInt(strNum.toString())

        return _prefix
    }
}